### ISSUE TITLE
Make our rich-default repo2docker compatible

### ISSUE THREAD

**Issue by rkevin-arch** [LibreTexts/default-env#45]
This would help manage our honestly very bloated Dockerfile since repo2docker would automatically use the `environment.yaml` and other files like `apt.txt` to install packages for us so we don't have to write gigantic Dockerfiles. This also gives us an excuse to maybe remove some unused packages, but we need to double check what can be removed. Finally, this helps if we also want to use it for the binder environment in ckeditor-binder-plugin, so we don't need shim Dockerfiles and can just have one repo for everything.

**Comment by moorepants** [comment]
Not sure why we didn't do this in the first place...

**Comment by rkevin-arch** [comment]
I see there's an attempt in the repo2docker folder, but Celine said it was a failed attempt. I'll look into it and see how I can make it work.

**Comment by moorepants** [comment]
Try this set of top level packages:

```
  - astropy=4.0.1.post1
  - beautifulsoup4=4.9.0
  - biopandas=0.2.5
  - bokeh=2.0.1
  - boost=1.63.0
  - boost-cpp=1.72.0
  - bqplot=0.12.6
  - cartopy=0.17.0
  - cloudpickle=1.3.0
  - cycler=0.10.0
  - cyipopt=0.1.9
  - cython=0.29.16
  - cytoolz=0.10.1
  - dask=2.14.0
  - dask-core=2.14.0
  - dill=0.3.1.1
  - fftw=3.3.8
  - future=0.18.2
  - gdal=3.0.4
  - gmt=6.0.0
  - h5py=2.10.0
  - hdf4=4.2.13
  - hdf5=1.10.5
  - ipopt=3.13.1
  - ipydatawidgets=4.0.1
  - ipykernel=5.2.0
  - ipyleaflet=0.12.3
  - ipympl=0.4.1
  - ipyparallel=6.2.4
  - ipyscales=0.4.0
  - ipython=7.13.0
  - ipywidgets=7.5.1
  - joblib=0.14.1
  - jupyter-rsession-proxy=1.1
  - jupyter-server-proxy=1.3.2
  - jupyter_client=6.1.3
  - jupyter_core=4.6.3
  - jupyterhub=1.0.0
  - jupyterlab=1.2.10
  - jupyterlab_server=1.1.1
  - llvmlite=0.31.0
  - make=4.3
  - matplotlib=3.1.2
  - matplotlib-base=3.1.2
  - mordred=1.2.0
  - mpmath=1.1.0
  - nb_conda_kernels=2.2.3
  - nbconvert=5.4.1
  - nbformat=5.0.4
  - networkx=2.4
  - nodejs=11.14.0
  - notebook=6.0.0
  - numba=0.48.0
  - numexpr=2.7.1
  - numpy=1.18.1
  - octave_kernel=0.31.1
  - opty=1.0.0
  - pandas=1.0.3
  - pandoc=2.9.2.1
  - patsy=0.5.1
  - pillow=7.1.1
  - pip=20.0.2
  - pydy=0.5.0
  - pygments=2.6.1
  - pygmt=0.1.0
  - pymol=1.9.0.0
  - pypdb=1.300
  - pyproj=2.6.0
  - pyshtools=4.6.2
  - python=3.6.10
  - pythreejs=2.2.0
  - pywavelets=1.1.1
  - pyyaml=5.3.1
  - pyzmq=19.0.0
  - r-askpass=1.1
  - r-assertthat=0.2.1
  - r-backports=1.1.6
  - r-base=3.6.1
  - r-base64enc=0.1_3
  - r-bit=1.1_15.2
  - r-bit64=0.9_7
  - r-bitops=1.0_6
  - r-blob=1.2.1
  - r-brew=1.0_6
  - r-broom=0.5.5
  - r-callr=3.4.3
  - r-caret=6.0_86
  - r-cellranger=1.1.0
  - r-class=7.3_16
  - r-classint=0.4_3
  - r-cli=2.0.2
  - r-clipr=0.7.0
  - r-codetools=0.2_16
  - r-colorspace=1.4_1
  - r-commonmark=1.7
  - r-config=0.3
  - r-covr=3.5.0
  - r-crayon=1.3.4
  - r-crosstalk=1.1.0.1
  - r-curl=4.3
  - r-data.table=1.12.8
  - r-dbi=1.1.0
  - r-dbplyr=1.4.2
  - r-desc=1.2.0
  - r-devtools=2.3.0
  - r-digest=0.6.25
  - r-dplyr=0.8.5
  - r-dt=0.13
  - r-e1071=1.7_3
  - r-ellipsis=0.3.0
  - r-evaluate=0.14
  - r-fansi=0.4.1
  - r-farver=2.0.3
  - r-fastmap=1.0.1
  - r-forcats=0.5.0
  - r-foreach=1.5.0
  - r-forecast=8.12
  - r-forge=0.2.0
  - r-fracdiff=1.5_1
  - r-fs=1.4.1
  - r-generics=0.0.2
  - r-ggplot2=3.3.0
  - r-gh=1.1.0
  - r-git2r=0.26.1
  - r-glue=1.4.0
  - r-gower=0.2.1
  - r-gridextra=2.3
  - r-gtable=0.3.0
  - r-haven=2.2.0
  - r-hexbin=1.28.1
  - r-highr=0.8
  - r-hms=0.5.3
  - r-htmltools=0.4.0
  - r-htmlwidgets=1.5.1
  - r-httpuv=1.5.2
  - r-httr=1.4.1
  - r-ini=0.3.1
  - r-ipred=0.9_9
  - r-irdisplay=0.7
  - r-irkernel=1.1
  - r-isoband=0.2.1
  - r-iterators=1.0.12
  - r-jsonlite=1.6.1
  - r-kernsmooth=2.23_16
  - r-knitr=1.28
  - r-labeling=0.3
  - r-later=1.0.0
  - r-lattice=0.20_41
  - r-lava=1.6.7
  - r-lazyeval=0.2.2
  - r-leaflet=2.0.3
  - r-leaflet.providers=1.9.0
  - r-lifecycle=0.2.0
  - r-lmtest=0.9_37
  - r-lubridate=1.7.8
  - r-magrittr=1.5
  - r-markdown=1.1
  - r-mass=7.3_51.5
  - r-matrix=1.2_18
  - r-memoise=1.1.0
  - r-mgcv=1.8_31
  - r-mime=0.9
  - r-modelmetrics=1.2.2.2
  - r-modelr=0.1.6
  - r-munsell=0.5.0
  - r-nlme=3.1_147
  - r-nnet=7.3_13
  - r-numderiv=2016.8_1.1
  - r-nycflights13=1.0.1
  - r-openssl=1.4.1
  - r-pbdzmq=0.3_3
  - r-pillar=1.4.3
  - r-pkgbuild=1.0.6
  - r-pkgconfig=2.0.3
  - r-pkgload=1.0.2
  - r-plogr=0.2.0
  - r-plyr=1.8.6
  - r-png=0.1_7
  - r-praise=1.0.0
  - r-prettyunits=1.1.1
  - r-proc=1.16.2
  - r-processx=3.4.2
  - r-prodlim=2019.11.13
  - r-progress=1.2.2
  - r-promises=1.1.0
  - r-ps=1.3.2
  - r-purrr=0.3.3
  - r-quadprog=1.5_8
  - r-quantmod=0.4.17
  - r-r2d3=0.2.3
  - r-r6=2.4.1
  - r-randomforest=4.6_14
  - r-rappdirs=0.3.1
  - r-raster=3.0_12
  - r-rcmdcheck=1.3.3
  - r-rcolorbrewer=1.1_2
  - r-rcpp=1.0.4.6
  - r-rcpparmadillo=0.9.860.2.0
  - r-rcurl=1.98_1.1
  - r-readr=1.3.1
  - r-readxl=1.3.1
  - r-recipes=0.1.10
  - r-rematch=1.0.1
  - r-rematch2=2.1.1
  - r-remotes=2.1.1
  - r-repr=1.1.0
  - r-reprex=0.3.0
  - r-reshape2=1.4.4
  - r-rex=1.1.2
  - r-rlang=0.4.5
  - r-rmarkdown=2.1
  - r-roxygen2=7.1.0
  - r-rpart=4.1_15
  - r-rprojroot=1.3_2
  - r-rsqlite=2.2.0
  - r-rstudioapi=0.11
  - r-rversions=2.0.1
  - r-rvest=0.3.5
  - r-scales=1.1.0
  - r-selectr=0.4_2
  - r-sessioninfo=1.1.1
  - r-sf=0.9_2
  - r-shiny=1.4.0.2
  - r-shinydashboard=0.7.1
  - r-sourcetools=0.1.7
  - r-sp=1.4_1
  - r-sparklyr=1.1.0
  - r-squarem=2020.2
  - r-stringi=1.4.6
  - r-stringr=1.4.0
  - r-survival=3.1_12
  - r-sys=3.3
  - r-testthat=2.3.2
  - r-tibble=3.0.0
  - r-tidyr=1.0.2
  - r-tidyselect=1.0.0
  - r-tidyverse=1.3.0
  - r-timedate=3043.102
  - r-tinytex=0.21
  - r-tseries=0.10_47
  - r-ttr=0.23_6
  - r-units=0.6_6
  - r-urca=1.3_0
  - r-usethis=1.6.0
  - r-utf8=1.1.4
  - r-uuid=0.1_4
  - r-vctrs=0.2.4
  - r-viridis=0.5.1
  - r-viridislite=0.3.0
  - r-whisker=0.4
  - r-withr=2.1.2
  - r-xfun=0.13
  - r-xml2=1.3.1
  - r-xopen=1.0.0
  - r-xtable=1.8_4
  - r-xts=0.12_0
  - r-yaml=2.2.1
  - r-zeallot=0.1.0
  - r-zoo=1.8_7
  - rdkit=2017.09.1
  - requests=2.23.0
  - resonance=0.22.0
  - rpy2=3.1.0
  - ruamel_yaml=0.15.80
  - scikit-image=0.16.2
  - scikit-learn=0.22.2.post1
  - scipy=1.4.1
  - seaborn=0.10.0
  - setuptools=46.1.3
  - shapely=1.7.0
  - six=1.14.0
  - sqlalchemy=1.3.16
  - sqlite=3.30.1
  - statsmodels=0.11.1
  - sympy=1.5.1
  - toolz=0.10.0
  - udunits2=2.2.27.6
  - urllib3=1.25.8
  - vincent=0.4.4
  - wheel=0.34.2
  - widgetsnbextension=3.5.1
  - xarray=0.15.1
  - xeus=0.23.5
  - xeus-cling=0.8.1
  - xlrd=1.2.0
  - yaml=0.2.3
  - pip:
    - ipymol==0.5
    - pyqt5-sip==4.19.18
    - pyqtwebengine==5.12.1
```

**Comment by rkevin-arch** [comment]
We have finally "finalized" our conda environment (still subject to change). The `environment.yaml` is [here](https://github.com/LibreTexts/default-env/blob/repo2docker/repo2docker/environment.yml) and should have everything in rich default, save for a couple of exceptions.

Notable conda packages that we don't have in this environment (please let us know if it's fine to remove these):
- `jupyterhub` and dependencies (no idea why any user would run their own JupyterHub in our hub)
- `pymol`, `ipymol` and its dependencies (many issues such as [metalc#135](https://github.com/LibreTexts/metalc/issues/135) wants it gone anyway, and users who need it can just install it themselves)
- `r-rsqlite` and `r-rcurl` (sounds like dependencies that are no longer needed? I can't think of a reason why they're here, and no one depends on them. Let us know if you want these.)
- `opty` (conda package doesn't have python3.7 support, we probably should repackage it on conda-forge.)
- `rpy2` and dependencies (rpy2 doesn't support R 4.0)

Other packages not in this environment (not really notable and should not affect users in any way):
- `ipyscales` (old ipydatawidgets dependency, no longer needed)
- `pykdtree` (old cartopy dependency, no longer needed)
- `conda` and its dependencies (repo2docker packages conda in a separate environment, so it is still available)
- `nodejs` (already installed via apt)
- `tini` (used in jupyter's base image, not needed by repo2docker who uses a different init process)
- some GUI libraries that's only in https://github.com/LibreTexts/metalc/issues/47 (such as qt, pmw, many xorg packages, etc) that won't work in a notebook image, probably included in our image by accident
- `xorg-libxi` and dependencies (xorg package that won't work without an X server, no idea who needs it)

Please let us know if this `environment.yaml` is acceptable. We haven't started porting over all the `apt` packages that should be installed, but this should be a huge step towards a repo2docker compatible rich default.

**Comment by moorepants** [comment]
Great!

Some comments:

- pymol,ipymol. we need to provide some info on installing this for the users. the chemistry group at Uof Ark use this
- r-rsqulite and r-rcurl should stay, these are core utilities that users may need/want even though none of our packages depend on them. For example, libretexts users use rcurl to work with data files stored on libretexts.org
- opty: this is a package I develop, I'll update it for 3.7
- rpy2, that's unfortunate but i guess we have to drop.
- nodejs. I would think this is a dependency of at least one package we have installed, no?

Other than that, it sounds good. I'll look through the file now.

**Comment by moorepants** [comment]
I am checking with the chemistry people about pymol.

**Comment by moorepants** [comment]
I don't see a PR to this repo for the changes.

**Comment by rkevin-arch** [comment]
- pymol and friends: We will add documentation on how to create conda environments and install their own persistent packages.
- r-rsqlite and r-rcurl: Adding that now and rebuilding.
- opty: Great, thanks!
- nodejs: Apparantly no conda package depends on it.
- `I don't see a PR to this repo for the changes`: They're in the branch called `repo2docker`, and there is no PR since it's not ready yet (notably all the apt packages are missing, which includes octave packages, basic utilities like `nano` and `vi`, and so on). You can still test the conda packages in the `rkevin/r2dtest:0.3` docker image (doesn't have `r-rsqlite` and `r-rcurl` yet, that is built earlier).

**Comment by moorepants** [comment]
Sounds good. Let me know when the PR is ready and I can have a look then.

**Comment by rkevin-arch** [comment]
Also, can we remove `pyqtwebengine`? It seems only useful when building qt applications (which needs a graphical interface to run, and we aren't running an X server or anything). No one has requested it (no GitHub issues I can find, anyway) and this would allow us to remove another pip package.

**Comment by moorepants** [comment]
That's fine, as long as it isn't a dependency. Funny conicidence, I was just cleaning up my personal conda envs and removed that package!

**Comment by rkevin-arch** [comment]
Cool. Now every package listed in `environment.yaml` are in there for a reason, and it is hopefully a top level package list that makes sense for future maintainers. The image is being built, and once it's ready it will be put on the staging cluster for testing.

**Comment by moorepants** [comment]
Nice work!

**Comment by rkevin-arch** [comment]
We have pretty much finalized the image and it is ready for testing. What do you think about having two options on the production hub like this:
![image](https://user-images.githubusercontent.com/5837628/89970375-9292c800-dc0d-11ea-952a-d949e690d425.png)
Let us know if this wording is acceptable and we'll deploy this on production.

**Comment by moorepants** [comment]
LGTM!

**Comment by rkevin-arch** [comment]
Hub updated, these two options should be available. We'll monitor jupyterteam@ucdavis.edu and see how it goes, and if nothing breaks for people then we can consider this image mostly final.

**Comment by rkevin-arch** [comment]
One sec, found something that can be improved in the current image...

**Comment by rkevin-arch** [comment]
Done! The current image on the production hub should have everything available in our old image, except pymol/ipymol, rpy2, and opty(https://github.com/conda-forge/opty-feedstock/issues/13).

**Comment by moorepants** [comment]
Now that this is live, are you going to make a PR to this repo to add repo2docker? Or is it merged?

**Comment by sandertyu** [comment]
At the moment, the environment is [here](https://github.com/LibreTexts/default-env/tree/repo2docker/repo2docker), in the repo2docker branch. I think Kevin has said he wants to wait for more testing before we merge to master. Currently, the latest version is 0.7.1 on both production and live, and that image is located on Kevin's dockerhub. 

I am okay with merging to master, that way we can all more intuitively find it. The branch does not overwrite anything currently there other than the unused repo2docker directory.

**Comment by rkevin-arch** [comment]
#47. I would merge it after the texlive change.

**Comment by rkevin-arch** [comment]
Closed by #47
